### PR TITLE
Fix 19970 - Cast from 0 to ptr is not null at CTFE

### PR DESCRIPTION
**Pull request by Geod24**
That fix seems a bit too specific, but it solves the bug. New test cases welcome @EyalIO .

### CODE REVIEW

**Review by UplinkCoder — APPROVED** [review]
funky but no objections here.
Perhaps try to not use new.

File: src/dmd/ctfeexpr.d [hunk 1]
```diff
@@ -746,6 +746,17 @@ Expression getAggregateFromPointer(Expression e, dinteger_t* ofs)
             return se.e1;
         }
     }
+    if (auto ie = e.isIntegerExp())
+        if (ie.type.ty == Tpointer)
+        {
+            // It can be a `null` disguised as a cast, e.g. `cast(void*)0`.
+            if (ie.getInteger() == 0)
```
Comment (RazvanN7) [inline src/dmd/ctfeexpr.d:753]: Is there any reason not to conflate this `if` with the one above by using `&&` ?

File: src/dmd/ctfeexpr.d [hunk 2]
```diff
@@ -746,6 +746,15 @@ Expression getAggregateFromPointer(Expression e, dinteger_t* ofs)
             return se.e1;
         }
     }
+
+    // It can be a `null` disguised as a cast, e.g. `cast(void*)0`.
+    if (auto ie = e.isIntegerExp())
```
Comment (UplinkCoder) [inline src/dmd/ctfeexpr.d:751]: yikes 3 nests no braces ...

File: src/dmd/ctfeexpr.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/ctfeexpr.d]: Done

**Review by atilaneves — APPROVED** [review]

File: src/dmd/ctfeexpr.d (hunk 2 above)
Comment (PetarKirov) [reply src/dmd/ctfeexpr.d:751]: Welcome to Python @UplinkCoder :smile

**Pull request merged**